[Soh How Yean, Adriel] iP

leirdas opened the pull request:
# John Doe
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

John Doe frees your mind from having to remember things you need to do. It's
- text-based
- easy to learn
- SUPER FAST to use

All you need to do is,
1. Download and run the program from [here](https://github.com/leirdas/ip)
2. Add your tasks
3. Let it manage your tasks for you

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events  

If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

leirdas committed:
Level 1. Greet, Echo, Exit

leirdas committed:
Level 2. Add, List

leirdas committed:
Level 3. Mark as Done

leirdas committed:
Level 4. ToDos, Events, Deadlines

leirdas committed:
Automated Text UI Testing

leirdas committed:
Level 5. Handle Errors

leirdas committed:
Level 6. Delete

leirdas committed:
Add use enums

leirdas committed:
Add Level 7. Save

leirdas committed:
Add Level 8. Dates and Times

leirdas committed:
Merge branch 'branch-Level-7'

leirdas committed:
Merge branch 'branch-Level-8'

leirdas committed:
Refactor code to extract out classes for more OOP

leirdas committed:
Divide class into packages

leirdas committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

leirdas committed:
Add JUnit tests

leirdas committed:
Package app as JAR file

leirdas committed:
Add JavaDoc comments

leirdas committed:
Refactor code to comply with coding standard

leirdas committed:
Merge branch 'branch-A-JavaDoc'

leirdas committed:
Merge branch 'branch-A-CodingStandard'

leirdas committed:
Add Level 9. Find

leirdas committed:
Merge branch 'branch-Level-9'

leirdas committed:
Refactor code

leirdas committed:
Refactor code

NicholasTYD commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,35 @@
+ * Command runs certain commands input by user.
+ */
+public abstract class Command {
+    boolean isExit;
```
```suggestion
    boolean isExitCommand;
```
Perhaps it might be better to use isExitCommand as a variable name instead of isExit, so that it is more intuitive when referring to this boolean in other classes?

NicholasTYD commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+     * @return Command to carry out.
+     * @throws DukeException
+     */
+    public static Command parse(String fullCommand) throws DukeException {
```
I like the way you formatted this method. The indentations and organization of your switch cases is really neat and clean.

NicholasTYD commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,155 @@
+package duke;
+
+import duke.command.*;
```
I feel that it may be better to list every imported class explicitly (and also because the coding standard said so).

NicholasTYD commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Constructor for the Duke chatbot.
+     * @param filePath filepath to save the tasks on.
+     */
```
```suggestion
    /**
     * Constructor for the Duke chatbot.
     *
     * @param filePath filepath to save the tasks on.
     */
```
I think that it may be better to leave an empty line between the description and parameter section so that it looks more organized and neat (Its part of the coding standard).

NicholasTYD commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,30 @@
+import duke.Ui;
+
+/**
+ * UnmarkCommand unmakrs a task as done.
```
```suggestion
 * UnmarkCommand unmarks a task as done.
```
I spotted a typo in the word "unmark". Perhaps you might want to rectify it?

NicholasTYD reviewed (commented):
Save for a few places where there are some really minor code violations, your code overall looks pretty organized and neat! Good job! If you have extra time, it would be good if you can fix the spacing on your Javadocs.

leirdas committed:
Add Level 10. GUI

leirdas committed:
Merge branch 'branch-Level-10'

leirdas committed:
Refactor code

jacobkwan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,152 @@
+
+    /**
+     * Returns an ArrayList containing task objects that were previously saved.
+     * @return ArrayList containing task objects.
```
I think it would be better to leave a line between the description and return section (as mentioned in the course coding standards)

jacobkwan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,152 @@
+     * @throws IOException
+     */
+    public ArrayList<Task> load() throws IOException {
+        if (!Files.exists(relativeDirectoryPath)) {
+            Files.createDirectory(relativeDirectoryPath);
+        }
+        if (!Files.exists(relativeFilePath)) {
+            Files.createFile(relativeFilePath);
+        }
```
I like the neat logic here, it's very clear what cases your code is defending against.

jacobkwan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,152 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+        Scanner sc = new Scanner(new File(String.valueOf(relativeFilePath)));
+        ArrayList<String> stringTasks = new ArrayList<>();
+        while(sc.hasNextLine()) {
```
I think there should be a space between `while` and the `(`?

jacobkwan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,152 @@
+                task = parseStringToTask(taskDetails[1], "E");
+                break;
+            }
+            default:
+                break;
```
I may be wrong, but perhaps it would be better to remove the default case if you are not doing anything useful inside. I understand the defensive logic with the `task != null` check but during the loading of the text file, if for some reason you were to read a `taskType` that was not `T`, `D` or `E`, would it be better to remove that line from the text file to avoid persisting invalid data?

jacobkwan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,152 @@
+            ArrayList<String> stringTasks = new ArrayList<>();
+            for (Task task : tasks) {
+                switch (task.getTaskType()) {
+                    case "T": {
+                        String data = "T" + " | " + task.getStatus() + " | " + task.getDescription();
+                        stringTasks.add(data);
+                        break;
+                    }
+                    case "D": {
+                        String data = "D" + " | " + task.getStatus() + " | " + task.getDescription();
+                        stringTasks.add(data);
+                        break;
+                    }
+                    case "E": {
+                        String data = "E" + " | " + task.getStatus() + " | " + task.getDescription();
+                        stringTasks.add(data);
+                        break;
+                    }
+                    default:
+                        break;
```
I believe the indentation for `case` statements should be in line with the `switch` line. Not sure why the indentation is correct in `load()` but not here 🤔, perhaps consider using a code formatter and enforcing this rule?

jacobkwan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    public Duke() {
+        ui = new Ui();
+        storage = new Storage(System.getProperty("user.home"));
```
Interesting use of system properties!

jacobkwan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+public class Parser {
+    private final static String INDENTATION = "   ";
+
+    private enum Commands {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND
+    }
```
Nice use of enums here, perhaps you could consider abstracting them out to a separate file? This might be useful if you ever needed them elsewhere in your codebase.

jacobkwan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+                case LIST:
+                    return new ListCommand();
+                case MARK:
+                    throw new DukeException(INDENTATION +
```
I may be wrong but so far I've seen all of your `DukeException`s prefixed with `INDENTATION` (which is just whitespace(?) ). Perhaps you could abstract this padding directly into the constructor of `DukeException`? Again, on the surface, to me it seems like a good piece to abstract out.

jacobkwan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,84 @@
+     * Constructor for duke.TaskList, which stores task entered by user in duke.Duke.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<Task>();
```
you could consider removing the `this` modifier since the variable `tasks` is not being shadowed (as advised by the course coding standard). This is present in a few places in the codebase too.

jacobkwan commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,162 @@
+     * Prints welcome message.
+     */
+    public void showWelcome() {
+        System.out.println(INDENTATION + LINE);
```
Similar to a previous comment, if `INDENTATION` is used every time you print to sys out, you could consider defining your own `print` method that takes in a String and adds the `INDENTATION` before printing.

jacobkwan commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,42 @@
+ * DeleteCommand deletes a task.
+ */
+public class DeleteCommand extends Command {
+    private int indexToDelete;
```
nice and intuitive variable naming!

jacobkwan commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    public ExitCommand() {
+        super();
+        this.isExit = true;
```
Not sure if I missed it, but `Duke.java` does not seem to be checking the value of `isExit` after `command.execute()`. So is there a purpose for setting `isExit = true`?

jacobkwan commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,40 @@
+        if (numOfTasks == 0) {
+            return "There are no matching tasks in your list.";
+        } else {
+            StringBuilder str = new StringBuilder();
+            str.append(String.format("Here %s the task%s in your list:\n",
+                    numOfTasks > 1 ? "are" : "is", numOfTasks == 1 ? "" : "s"));
+            for (int i = 1; i <= numOfTasks; i++) {
+                str.append(String.format("%s. %s\n", String.valueOf(i), foundTasks.getTask(i).toString()));
+            }
+            return str.toString();
```
perhaps you could consider abstracting away part of the job of string creation to `TaskList` as it could be useful in `ListCommand` as well? Try finding the generic representation that could be reused multiple times.

jacobkwan commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Constructor for Event.
+     * @param description Event description.
+     * @param at when the event is happening.
```
I believe `when` should be capitalised to `When` as mentioned in the coding standard. Occurs in other places too.

jacobkwan reviewed (commented):
LGTM, good abstractions in general, just some minor nits to fix 👍

leirdas committed:
Add checkstyle

leirdas committed:
Merge branch 'branch-A-CheckStyle'

leirdas committed:
Generate new JAR file

Use gradle to generate new JAR file to run duke

leirdas committed:
Add assertions

Use assert feature to document important assumptions that should hold
at various points in the code.

leirdas committed:
Merge pull request #1 from leirdas/branch-A-Assertions

Add assertions

leirdas committed:
Improve code quality

There were complicated expressions in the code and large sections of
code that might not be related were grouped together.

Improving the code quality is important as it makes it easier for code
to be read, understood and modified by other developers later on.

Let's,
* change complicated expressions to more readable code
* group related block of code together

leirdas committed:
Merge branch 'master' into branch-A-CodeQuality

leirdas committed:
Merge pull request #2 from leirdas/branch-A-CodeQuality

Improve code quality

leirdas committed:
Merge pull request #3 from leirdas/master

Merge pull request #2 from leirdas/branch-A-CodeQuality

leirdas committed:
Set up CI

Use GitHub Actions to set up Continuous Integration.

leirdas committed:
Merge branch 'branch-A-CodeQuality'

leirdas committed:
Add priority enum to Task class

leirdas committed:
Add a way to attach priorities to task

leirdas committed:
Merge branch 'branch-C-Priority'

leirdas committed:
Fix errors in loading tasks

leirdas committed:
Make changes to GUI

leirdas committed:
Fix code violating checkstyles

leirdas committed:
Add Ui.png

leirdas committed:
Update README.md

leirdas committed:
Tweak code

leirdas committed:
Update README.md
